File: 0000-0999/0025.ReverseNodesink-Group.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseKGroup(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        if not head or k <= 1:
            return head
        p = head
        for i in xrange(k):
            if not p:
                return head
            p = p.next
        pre = self.reverseKGroup(p, k)
        nxt = head.next
        for i in xrange(k-1):
            head.next = pre
            pre = head
            head = nxt
            nxt = head.next
        head.next = pre
        return head

File: 0000-0999/0025.ReverseNodesink-Group.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func reverseKGroup(head *ListNode, k int) *ListNode {
    if head == nil || k <= 1 {
        return head
    }
    p := head
    for i := 0; i < k; i++ {
        if p == nil {
            return head
        }
        p = p.Next
    }
    newHead := reverseKGroup(p, k)
    for i := 0; i < k; i++ {
        nxt := head.Next
        head.Next = newHead
        newHead = head
        head = nxt
    }
    return newHead
}